README
README file for Image::Magick::Chart.

See also: Changes.txt.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip Image-Magick-Chart-1.00.tgz
shell>tar mxvf Image-Magick-Chart-1.00.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run Chart.pm through you favourite pod2html translator.

If you are using my fancy-pom2.pl, with its 'default.css' file installed in
/apache2/htdocs/css/, you'd do:

shell>perl fancy-pom2.pl html -css Chart.pm > /apache2/htdocs/Chart.html

or perhaps something like:

shell>perl fancy-pom2.pl html -css Chart.pm > /perl/html/site/lib/Image/Magick/Chart.html

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip Image-Magick-Chart-1.00.zip
shell>ppm install --location=. Image-Magick-Chart
shell>del Image-Magick-Chart-1.00.ppd
shell>del PPM-Image-Magick-Chart-1.00.tar.gz

CHANGES
Revision history for Perl extension Image::Magick::Chart.

1.07  Unreleased
	- Replace /usr/bin/perl with /usr/bin/env perl.
	- Replace common::sense with use strict and use warnings, to get uninit var warnings.

1.06  Sun Feb 21 12:55:21 2010
	- Remove text 'All rights reserved' (for Debian licensing).
	- Remove POD heads 'Required Modules' and 'Changes'.

1.05  Wed Feb 10 14:02:08 2010
	- MANIFEST.SKIP updated to exclude MYMETA.yml. Add MYMETA.yml.

1.04  Fri Nov 13 13:20:00 2009
	- Run dos2unix
	- Rename Changes.txt to CHANGES

1.03  Fri Dec 12 18:15:00 2008
	  - Image::Magick's Draw() method has an option, 'fill'. When I first wrote this module,
	  	fill's value defaulted to 'none', and now it defaults to something else. The effect is
		that when drawing a frame, the whole image is now filled in with black. Hence the patch
		to Image::Magick::Chart's draw_frame() method.
		I am tempted to rant and rave about the endless butchering of Image::Magick by its author,
		but that would not achieve anything... And anyway, it's an extremely useful module

1.02  Sun Jul 17 17:52:00 2005
	- Add an option, y_axis_labels_x, which allows the user to specify the x-axis value
		(abscissa) at which to start writing y-axis labels. From the POD:
		The value, if not undef, determines the x-axis value (abscissa) at which y-axis labels are written.
		The special value undef means this module calculates an abscissa at which to start writing
		y-axis labels. The default behaviour is identical to the previous version 1.01.
		This calculation will only produce a pretty-looking column of y-axis labels when all labels are
		the same width in pixels. See C<sub draw_y_axis_labels()> for the calculation.
		The default value is undef.
		This parameter is optional.

1.01  Mon Jun 13 12:19:00 2005
	- No code changes
	- Various documentation fixes and clarifications

1.00  Thu May 27 11:25:29 2005
	- original version


examples/test-chart.pl
#!/usr/bin/perl

use strict;
use warnings;

use Image::Magick::Chart::HorizontalBars;

# -----------------------

Image::Magick::Chart::HorizontalBars -> new
(
	antialias				=> 0,	# 0 => No antialias; 1 => Antialias.
	bar_width				=> 8,	# Pixels.
	bg_color				=> 'white',
	colorspace				=> 'RGB',
	depth					=> 8,	# Bits per channel.
	fg_color				=> 'blue',
	font					=> 'Courier',
	frame_color				=> 'black',
	frame_option			=> 1,	# 0 => None; 1 => Draw it.
	height					=> 0,
	image					=> '',
	output_file_name		=> 'image-1.png',
	padding					=> [30, 30, 30, 30],	# [12 noon, 3, 6, 9].
	pointsize				=> 14,	# Points.
	tick_length				=> 4,	# Pixels.
	title					=> 'Percent (%)',
	width					=> 0,
	x_axis_data				=> [0, 20, 40, 60, 80, 100],
	x_axis_labels			=> [0, 20, 40, 60, 80, 100],
	x_axis_labels_option	=> 1,	# 0 => None; 1 => Draw them.
	x_axis_ticks_option		=> 2,	# 0 => None; 1 => Below x-axis; 2 => Across frame.
	x_data					=> [15, 5, 70, 25, 45, 20, 65],
	x_data_option			=> 1,
	x_pixels_per_unit		=> 3,	# Horizontal width of each data unit.
	y_axis_data				=> [1 .. 7, 8], # 7 data points, plus 1 to make result pretty.
	y_axis_labels			=> [(map{"($_)"} reverse (1 .. 7) ), ''],
	y_axis_labels_option	=> 1,	# 0 => None; 1 => Draw them.
	y_axis_ticks_option		=> 1,	# 0 => None; 1 => Left of y-axis; 2 => Across frame.
	y_pixels_per_unit		=> 20,
) -> draw();

Image::Magick::Chart::HorizontalBars -> new
(
	antialias				=> 0,	# 0 => No antialias; 1 => Antialias.
	bar_width				=> 8,	# Pixels.
	bg_color				=> 'white',
	colorspace				=> 'RGB',
	depth					=> 8,	# Bits per channel.
	fg_color				=> 'blue',
	font					=> 'Courier',
	frame_color				=> 'black',
	frame_option			=> 1,	# 0 => None; 1 => Draw it.
	height					=> 0,
	image					=> '',
	output_file_name		=> 'image-2.png',
	padding					=> [30, 30, 30, 30],	# [12 noon, 3, 6, 9].
	pointsize				=> 14,	# Points.
	tick_length				=> 4,	# Pixels.
	title					=> 'Mean',
	width					=> 0,
	x_axis_data				=> [0 .. 5],
	x_axis_labels			=> [0 .. 5],
	x_axis_labels_option	=> 1,	# 0 => None; 1 => Draw them.
	x_axis_ticks_option		=> 2,	# 0 => None; 1 => Below x-axis; 2 => Across frame.
	x_data					=> [4.0, 3.5, 4.0, 4.6, 3.9, 4.0, 3.0, 3.0, 3.8, 4.0, 3.5],
	x_data_option			=> 1,
	x_pixels_per_unit		=> 60,	# Horizontal width of each data unit.
	y_axis_data				=> [1 .. 11, 12], # 11 data points, plus 1 to make result pretty.
	y_axis_labels			=> [reverse (1 .. 11), ''],
	y_axis_labels_option	=> 1,	# 0 => None; 1 => Draw them.
	y_axis_ticks_option		=> 1,	# 0 => None; 1 => Left of y-axis; 2 => Across frame.
	y_pixels_per_unit		=> 20,
) -> draw();


lib/Image/Magick/Chart/HorizontalBars.pm
package Image::Magick::Chart::HorizontalBars;

# Name:
#	Image::Magick::Slice.
#
# Documentation:
#	POD-style documentation is at the end. Extract it with pod2html.*.
#
# Reference:
#	Object Oriented Perl
#	Damian Conway
#	Manning
#	1-884777-79-1
#	P 114
#
# Note:
#	o Tab = 4 spaces || die.
#
# Author:
#	Ron Savage <ron@savage.net.au>
#	Home page: http://savage.net.au/index.html
#
# Licence:
#	Australian copyright (c) 2005 Ron Savage.
#
#	All Programs of mine are 'OSI Certified Open Source Software';
#	you can redistribute them and/or modify them under the terms of
#	The Artistic License, a copy of which is available at:
#	http://www.opensource.org/licenses/index.html

use strict;
use warnings;

use Carp;
use Image::Magick::Chart;

require 5.005_62;

require Exporter;

our @ISA = qw(Image::Magick::Chart);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Image::Magick::Chart::HorizontalBars ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(

) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);
our $VERSION = '1.06';

# -----------------------------------------------

# Preloaded methods go here.

# -----------------------------------------------

# Encapsulated class data.

{
	my(%_attr_data) =
	(
	);

	sub _default_for
	{
		my($self, $attr_name) = @_;

		$_attr_data{$attr_name};
	}

	sub _standard_keys
	{
		keys %_attr_data;
	}

}	# End of encapsulated class data.

# -----------------------------------------------

sub draw
{
	my($self) = @_;

	$self -> draw_frame()			if ($$self{'_frame_option'});
	$self -> draw_x_axis_ticks()	if ($$self{'_x_axis_ticks_option'});
	$self -> draw_x_axis_labels()	if ($$self{'_x_axis_labels_option'});
	$self -> draw_y_axis_ticks()	if ($$self{'_y_axis_ticks_option'});
	$self -> draw_y_axis_labels()	if ($$self{'_y_axis_labels_option'});
	$self -> draw_horizontal_bars();
	$self -> draw_title()			if ($$self{'_title'});
	$self -> write();

}	# End of draw.

# -----------------------------------------------

sub new
{
	my($class, %arg)	= @_;
	my($self)			= Image::Magick::Chart -> new(%arg);
	$self				= bless($self, $class); # Reconsecrate.

	return $self;

}	# End of new.

# -----------------------------------------------

1;

__END__

=head1 NAME

C<Image::Magick::Chart::HorizontalBars> - Use Image::Magick to create charts.

=head1 Synopsis

	#!/usr/bin/env perl

	use Image::Magick::Chart::HorizontalBars;

	Image::Magick::Chart::HorizontalBars -> new
	(
		antialias            => 0, # 0 => No antialias; 1 => Antialias.
		bar_width            => 8, # Pixels.
		bg_color             => 'white',
		colorspace           => 'RGB',
		depth                => 8, # Bits per channel.
		fg_color             => 'blue',
		font                 => 'Courier',
		frame_color          => 'black',
		frame_option         => 1, # 0 => None; 1 => Draw it.
		height               => 0,
		image                => '',
		output_file_name     => 'image-1.png',
		padding              => [30, 30, 30, 30], # [12 noon, 3, 6, 9].
		pointsize            => 14, # Points.
		tick_length          => 4,  # Pixels.
		title                => 'Percent (%)',
		width                => 0,
		x_axis_data          => [0, 20, 40, 60, 80, 100],
		x_axis_labels        => [0, 20, 40, 60, 80, 100],
		x_axis_labels_option => 1, # 0 => None; 1 => Draw them.
		x_axis_ticks_option  => 2, # 0 => None; 1 => Below x-axis; 2 => Across frame.
		x_data               => [15, 5, 70, 25, 45, 20, 65],
		x_data_option        => 1,
		x_pixels_per_unit    => 3, # Horizontal width of each data unit.
		y_axis_data          => [1 .. 7, 8], # 7 data points, plus 1 to make image pretty.
		y_axis_labels        => [(map{"($_)"} reverse (1 .. 7) ), ''],
		y_axis_labels_option => 1, # 0 => None; 1 => Draw them.
		y_axis_ticks_option  => 1, # 0 => None; 1 => Left of y-axis; 2 => Across frame.
		y_pixels_per_unit    => 20,
	) -> draw();

This code is part of examples/test-chart.pl.

Note: You do not need to specify all the options above, of course, but only those you
wish to differ from the defaults. I've included all options in examples/test-chart.pl
just to save you the effort of having to type them in.

=head1 Description

C<Image::Magick::Chart::HorizontalBars> is a pure Perl module.

This module uses Image::Magick to create simple charts (images consisting of horizontal bars)
with optional axes, axis labels, etc.

See examples/image-*.png for sample output, and examples/test-chart.pl for the program
which created those samples.

See the docs for C<Image::Magick::Chart> for details.

=head1 Distributions

This module is available both as a Unix-style distro (*.tgz) and an
ActiveState-style distro (*.ppd). The latter is shipped in a *.zip file.

See http://savage.net.au/Perl-modules.html for details.

See http://savage.net.au/Perl-modules/html/installing-a-module.html for
help on unpacking and installing each type of distro.

=head1 Constructor and initialization

new(...) returns a C<Image::Magick::Chart::HorizontalBars> object.

This is the class's contructor.

Usage: Image::Magick::Chart::HorizontalBars -> new().

This method takes a set of parameters.

For each parameter you wish to use, call new as new(param_1 => value_1, ...).

Any parameter which is supported by the parent class, Image::Magick::Chart, can be used
in the call the C<new()> in this class. So, see the docs for Image::Magick::Chart.

=head1 Method: draw()

Output the chart.

=head1 Method: new(...)

Returns a object of type C<Image::Magick::Chart::HorizontalBars>.

See above, in the section called 'Constructor and initialization' for details.

=head1 Author

C<Image::Magick::Chart::HorizontalBars> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2005.

Home page: http://savage.net.au/index.html

=head1 Copyright

Australian copyright (c) 2005, Ron Savage.
	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut
